File: atk-adaptor/hyperlink.c
/*
 * AT-SPI - Assistive Technology Service Provider Interface
 * (Gnome Accessibility Project; http://developer.gnome.org/projects/gap)
 *
 * Copyright 2008 Novell, Inc.
 * Copyright 2001, 2002 Sun Microsystems Inc.,
 * Copyright 2001, 2002 Ximian, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include "accessible.h"

static AtkHyperlink *
get_hyperlink (DBusMessage * message)
{
  AtkObject *obj = spi_dbus_get_object (dbus_message_get_path (message));
  if (!obj)
    return NULL;
  return ATK_HYPERLINK (obj);
}

static AtkHyperlink *
get_hyperlink_from_path (const char *path, void *user_data)
{
  AtkObject *obj = spi_dbus_get_object (path);
  if (!obj || !ATK_IS_HYPERLINK(obj))
    return NULL;
  return ATK_HYPERLINK (obj);
}

static dbus_bool_t
impl_get_nAnchors (const char *path, DBusMessageIter * iter, void *user_data)
{
  AtkHyperlink *link = get_hyperlink_from_path (path, user_data);
  if (!link)
    return FALSE;
  return droute_return_v_int32 (iter, atk_hyperlink_get_n_anchors (link));
}


static dbus_bool_t
impl_get_startIndex (const char *path, DBusMessageIter * iter,
		     void *user_data)
{
  AtkHyperlink *link = get_hyperlink_from_path (path, user_data);
  if (!link)
    return FALSE;
  return droute_return_v_int32 (iter, atk_hyperlink_get_start_index (link));
}

static dbus_bool_t
impl_get_endIndex (const char *path, DBusMessageIter * iter, void *user_data)
{
  AtkHyperlink *link = get_hyperlink_from_path (path, user_data);
  if (!link)
    return FALSE;
  return droute_return_v_int32 (iter, atk_hyperlink_get_end_index (link));
}

static DBusMessage *
impl_getObject (DBusConnection * bus, DBusMessage * message, void *user_data)
{
  AtkHyperlink *link = get_hyperlink (message);
  DBusError error;
  dbus_int32_t i;
  AtkObject *atk_object;

  if (!link)
    return spi_dbus_general_error (message);
  dbus_error_init (&error);
  if (!dbus_message_get_args
      (message, &error, DBUS_TYPE_INT32, &i, DBUS_TYPE_INVALID))
    {
      return SPI_DBUS_RETURN_ERROR (message, &error);
    }
  atk_object = atk_hyperlink_get_object (link, i);
  return spi_dbus_return_object (message, atk_object, FALSE);
}

static DBusMessage *
impl_getURI (DBusConnection * bus, DBusMessage * message, void *user_data)
{
  AtkHyperlink *link = get_hyperlink (message);
  dbus_int32_t i;
  DBusError error;
  gchar *rv;
  DBusMessage *reply;

  if (!link)
    return spi_dbus_general_error (message);
  dbus_error_init (&error);
  if (!dbus_message_get_args
      (message, &error, DBUS_TYPE_INT32, &i, DBUS_TYPE_INT32, &i,
       DBUS_TYPE_INVALID))
    {
      return SPI_DBUS_RETURN_ERROR (message, &error);
    }

  rv = atk_hyperlink_get_uri (link, i);
  if (!rv)
    rv = g_strdup ("");
  reply = dbus_message_new_method_return (message);
  if (reply)
    {
      dbus_message_append_args (reply, DBUS_TYPE_STRING, &rv,
				DBUS_TYPE_INVALID);
    }
  g_free (rv);
  return reply;
}

static DBusMessage *
impl_isValid (DBusConnection * bus, DBusMessage * message, void *user_data)
{
  AtkHyperlink *link = get_hyperlink (message);
  dbus_bool_t rv;
  DBusMessage *reply;

  if (!link)
    return spi_dbus_general_error (message);

  rv = atk_hyperlink_is_valid (link);
  reply = dbus_message_new_method_return (message);
  if (reply)
    {
      dbus_message_append_args (reply, DBUS_TYPE_BOOLEAN, &rv,
				DBUS_TYPE_INVALID);
    }
  return reply;
}

static DRouteMethod methods[] = {
  {impl_getObject, "getObject"},
  {impl_getURI, "getURI"},
  {impl_isValid, "isValid"},
  {NULL, NULL}
};

static DRouteProperty properties[] = {
  {impl_get_nAnchors, NULL, "nAnchors"},
  {impl_get_startIndex, NULL, "startIndex"},
  {impl_get_endIndex, NULL, "endIndex"},
  {NULL, NULL, NULL}
};

void
spi_initialize_hyperlink (DRouteData * data)
{
  droute_add_interface (data, SPI_DBUS_INTERFACE_HYPERLINK,
			methods, properties,
			(DRouteGetDatumFunction) get_hyperlink_from_path,
			NULL);
};

File: pyatspi/hyperlink.py
#Copyright (C) 2008 Codethink Ltd

#This library is free software; you can redistribute it and/or
#modify it under the terms of the GNU Lesser General Public
#License version 2 as published by the Free Software Foundation.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#You should have received a copy of the GNU Lesser General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

from interfaces import *
from accessible import Accessible
from factory import accessible_factory

import dbus

__all__ = [
           "Hyperlink",
          ]

#------------------------------------------------------------------------------

class Hyperlink(Accessible):
        """
        Instances of Hyperlink are returned by Hypertext objects, and
        are the means by which end users and clients interact with linked,
        and in some cases embedded, content. Hyperlinks may have multiple
        "anchors", where an anchor corresponds to a reference to a particular
        resource with a corresponding resource identified (URI). Hyperlinks
        may be queried for their URIs, or queried for the objects corresponding
        to their anchors. The objects thus obtained are instances of
        Accessible, and may be queried, and manipulated via the Action
        interface.
        """

        def getObject(self, index):
                """
                Gets the i'th object, (where i is an integer between 0 and Hyperlink::numAnchors
                - 1, inclusive) associated with a Hyperlink. The objects returned
                are usually actionable (i.e. they should implement Accessibility::Action),
                and the available actions often include "open", "bookmark", "save
                link as", etc. They may also implement Accessibility::StreamableContent,
                although clients can normally use getURI to obtain a resource
                locator via which the object's data may be accessed.
                @return an Accessible object instance representing the Hyperlink's
                ith anchor, or through which the content associated with the
                ith anchor can be accessed.
                """
                func = self.get_dbus_method("getObject", dbus_interface=ATSPI_HYPERLINK)
                return func(index)

        def getURI(self, index):
                """
                Obtain a resource locator ('URI') which can be used to access
                the content to which this link "points" or is connected. 
                @return a string corresponding to the URI of the Hyperlink's
                'ith' anchor, if one exists, or a NIL string otherwise.
                """
                func = self.get_dbus_method("getURI", dbus_interface=ATSPI_HYPERLINK)
                return func(index)

        def isValid(self):
                """
                Check the hyperlink to see if a connection to its backing content
                can be established, or if its URI is valid. 
                @return True if the object's content is available, or False if
                the hyperlink's URI is invalid, or a connection to the resource
                can not be established.
                """
                func = self.get_dbus_method("isValid", dbus_interface=ATSPI_HYPERLINK)
                return func()

        def get_endIndex(self):
                return dbus.Int32(self._pgetter(self._dbus_interface, "endIndex"))
        _endIndexDoc = \
                """
                the ending offset within the containing Hypertext content with
                which this Hyperlink is associated; that is, the offset of the
                first element past the range within the Hypertext associated
                with this Hyperlink.
                """
        endIndex = property(fget=get_endIndex, doc=_endIndexDoc)

        def get_nAnchors(self):
                return dbus.Int16(self._pgetter(self._dbus_interface, "nAnchors"))
        _nAnchorsDoc = \
                """
                the number of separate anchors associated with this Hyperlink
                """
        nAnchors = property(fget=get_nAnchors, doc=_nAnchorsDoc)

        def get_startIndex(self):
                return dbus.Int32(self._pgetter(self._dbus_interface, "startIndex"))
        _startIndexDoc = \
                """
                the starting offset within the containing Hypertext content with
                which this Hyperlink is associated
                """
        startIndex = property(fget=get_startIndex, doc=_startIndexDoc)

# Register the accessible class with the factory.
accessible_factory.register_accessible_class(ATSPI_HYPERLINK, Hyperlink)

#END----------------------------------------------------------------------------
